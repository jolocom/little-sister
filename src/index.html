<!DOCTYPE html>

<html>
  <head>
    <meta charset="UTF-8" />
    <meta name="viewport"
          content="width=device-width, initial-scale=1, user-scalable=no">

    <title>Little Sister</title>

    <link href="https://fonts.googleapis.com/css?family=Roboto:100,400,300,600"
          rel="stylesheet"
          type="text/css" />
    <link href="https://fonts.googleapis.com/icon?family=Material+Icons"
          rel="stylesheet">

    <style type="text/css">
      * {
        margin: 0;
        padding: 0;
      }
      html, body {
        height: 100%;
      }
      body {
        font-family: Roboto;
      }
      #app {
        width: 100%;
        height: 100%;
      }
    </style>

    <!-- <meta http-equiv="Content-Security-Policy" content="default-src * data: gap: https://ssl.gstatic.com 'unsafe-eval'; style-src * 'unsafe-inline'; media-src *; img-src * data:"> -->
    <meta name="format-detection" content="telephone=no">
    <meta name="msapplication-tap-highlight" content="no">
    <meta name="viewport" content="user-scalable=no, initial-scale=1, maximum-scale=1, minimum-scale=1, width=device-width">
  </head>
  <body>
    <div id="app"></div>
    <script src="https://solid.github.io/releases/rdflib.js/rdflib-0.7.0.min.js"></script>
    <script src="https://d3js.org/d3.v4.min.js"></script>
    <script src="js/bundle.js"></script>
  </body>
</html>
