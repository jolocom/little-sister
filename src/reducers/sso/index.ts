import { AnyAction } from 'redux'

export interface StateVerificationSummary {
  id: string
  issuer: string
  selfSigned: boolean
  expires: string | undefined | Date
}

export interface StateTypeSummary {
  type: string
  values: string[]
  verifications: StateVerificationSummary[]
}

export interface StateCredentialRequestSummary {
  readonly callbackURL: string
  readonly requester: string
  readonly availableCredentials: StateTypeSummary[]
  readonly requestJWT: string
}

export interface StatePaymentRequestSummary {
  receiver: {
    did: string
    address: string
  }
  callbackURL: string
  amount: number
  description: string
  paymentRequest: string
}

export interface StateAuthenticationRequestSummary {
  callbackURL: string
  requester: string
  description: string
  requestJWT: string
}

export interface SsoState {
  activeCredentialRequest: StateCredentialRequestSummary
  activePaymentRequest: StatePaymentRequestSummary
  activeAuthenticationRequest: StateAuthenticationRequestSummary
  deepLinkLoading: boolean
}

const initialState: SsoState = {
  activeCredentialRequest: {
    requester: '',
    callbackURL: '',
    availableCredentials: [],
    requestJWT: '',
  },
  activePaymentRequest: {
    receiver: {
      did: '',
      address: '',
    },
    callbackURL: '',
    amount: 0,
    description: '',
    paymentRequest: '',
  },
  // add blank authentication request, which is did, public profile?
  activeAuthenticationRequest: {
    requester: '',
    callbackURL: '',
    description: '',
    requestJWT: '',
  },
  deepLinkLoading: false,
}

export const ssoReducer = (
  state = initialState,
  action: AnyAction,
): SsoState => {
  switch (action.type) {
    case 'SET_CREDENTIAL_REQUEST':
      return { ...state, activeCredentialRequest: action.value }
    case 'SET_PAYMENT_REQUEST':
      return { ...state, activePaymentRequest: action.value }
    case 'SET_AUTHENTICATION_REQUEST':
      return { ...state, activeAuthenticationRequest: action.value }
    case 'SET_DEEP_LINK_LOADING':
      return { ...state, deepLinkLoading: action.value }
    case 'CLEAR_INTERACTION_REQUEST':
      return initialState
    default:
      return state
  }
}
