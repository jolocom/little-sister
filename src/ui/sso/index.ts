export { Consent } from 'src/ui/sso/containers/consent'
