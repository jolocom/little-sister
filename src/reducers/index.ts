import { combineReducers } from 'redux'
import { accountReducer, AccountState } from 'src/reducers/account/'
import {
  registrationReducer,
  RegistrationState,
} from 'src/reducers/registration/'
import { navigationReducer } from 'src/reducers/navigation/'
import { NavigationState } from 'react-navigation'
import { ssoReducer, SsoState } from 'src/reducers/sso/'

export const rootReducer = combineReducers({
  account: accountReducer,
  registration: registrationReducer,
  navigation: navigationReducer,
  sso: ssoReducer,
})

export interface RootState {
  readonly account: AccountState
  readonly registration: RegistrationState
  readonly navigation: NavigationState
  readonly sso: SsoState
}
