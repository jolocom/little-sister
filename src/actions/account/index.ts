import { AnyAction, Dispatch } from 'redux'
import { genericActions, navigationActions } from 'src/actions/'
import { BackendMiddleware } from 'src/backendMiddleware'
import { routeList } from 'src/routeList'
import { DecoratedClaims, CategorizedClaims } from 'src/reducers/account'
import { SignedCredential } from 'jolocom-lib/js/credentials/signedCredential/signedCredential'
import {
  getClaimMetadataByCredentialType,
  getCredentialUiCategory,
  getUiCredentialTypeByType,
  instantiateIdentityWallet,
} from '../../lib/util'
import { cancelReceiving } from '../sso'

export const setDid = (did: string) => ({
  type: 'DID_SET',
  value: did,
})

export const setSelected = (claim: DecoratedClaims) => ({
  type: 'SET_SELECTED',
  selected: claim,
})

export const resetSelected = () => ({
  type: 'RESET_SELECTED',
})

export const handleClaimInput = (fieldValue: string, fieldName: string) => ({
  type: 'HANLDE_CLAIM_INPUT',
  fieldName,
  fieldValue,
})

export const toggleClaimsLoading = (value: boolean) => ({
  type: 'TOGGLE_CLAIMS_LOADING',
  value,
})

export const checkIdentityExists = () => async (
  dispatch: Dispatch<AnyAction>,
  getState: Function,
  backendMiddleware: BackendMiddleware,
) => {
  const { storageLib } = backendMiddleware

  try {
    const personas = await storageLib.get.persona()
    if (!personas.length) {
      dispatch(toggleLoading(false))
      return
    }

    dispatch(setDid(personas[0].did))
    await instantiateIdentityWallet(backendMiddleware)

    dispatch(toggleLoading(false))
    dispatch(navigationActions.navigatorReset({ routeName: routeList.Home }))
  } catch (err) {
    if (err.message.indexOf('no such table') === 0) {
      return
    }
    dispatch(genericActions.showErrorScreen(err))
  }
}

export const setIdentityWallet = () => async (
  dispatch: Dispatch<AnyAction>,
  getState: Function,
  backendMiddleware: BackendMiddleware,
) => {
  try {
    await instantiateIdentityWallet(backendMiddleware)
  } catch (err) {
    dispatch(genericActions.showErrorScreen(err))
  }
}

export const openClaimDetails = (claim: DecoratedClaims) => (
  dispatch: Dispatch<AnyAction>,
) => {
  dispatch(setSelected(claim))
  dispatch(
    navigationActions.navigate({
      routeName: routeList.ClaimDetails,
    }),
  )
}

export const saveClaim = () => async (
  dispatch: Dispatch<AnyAction>,
  getState: Function,
  backendMiddleware: BackendMiddleware,
) => {
  const { identityWallet, storageLib, keyChainLib } = backendMiddleware

  try {
    const did = getState().account.did.get('did')
    const claimsItem = getState().account.claims.toJS().selected
    const password = await keyChainLib.getPassword()

    const verifiableCredential = await identityWallet.create.signedCredential(
      {
        metadata: getClaimMetadataByCredentialType(claimsItem.credentialType),
        claim: claimsItem.claimData,
        subject: did,
      },
      password,
    )

    if (claimsItem.id) {
      await storageLib.delete.verifiableCredential(claimsItem.id)
    }

    await storageLib.store.verifiableCredential(verifiableCredential)
    await setClaimsForDid()

    dispatch(
      navigationActions.navigatorReset({
        routeName: routeList.Home,
      }),
    )
  } catch (err) {
    dispatch(genericActions.showErrorScreen(err))
  }
}

// TODO Currently only rendering  / adding one
export const saveExternalCredentials = () => async (
  dispatch: Dispatch<AnyAction>,
  getState: Function,
  backendMiddleware: BackendMiddleware,
) => {
  const { storageLib } = backendMiddleware
  const externalCredentials = getState().account.claims.toJS().pendingExternal
  const cred: SignedCredential = externalCredentials[0]

  if (cred.id) {
    await storageLib.delete.verifiableCredential(cred.id)
  }

  try {
    await storageLib.store.verifiableCredential(externalCredentials[0])
    dispatch(cancelReceiving())
  } catch (err) {
    dispatch(genericActions.showErrorScreen(err))
  }
}

export const toggleLoading = (value: boolean) => ({
  type: 'SET_LOADING',
  value,
})

export const setClaimsForDid = () => async (
  dispatch: Dispatch<AnyAction>,
  getState: Function,
  backendMiddleware: BackendMiddleware,
) => {
  dispatch(toggleClaimsLoading(true))
  const storageLib = backendMiddleware.storageLib

  const verifiableCredentials: SignedCredential[] = await storageLib.get.verifiableCredential()
  const claims = prepareClaimsForState(
    verifiableCredentials,
  ) as CategorizedClaims

  dispatch({
    type: 'SET_CLAIMS_FOR_DID',
    claims,
  })

  dispatch(toggleClaimsLoading(false))
}

const prepareClaimsForState = (credentials: SignedCredential[]) => {
  const categorizedClaims = {}
  const decoratedCredentials = convertToDecoratedClaim(credentials)

  decoratedCredentials.forEach(decoratedCred => {
    const uiCategory = getCredentialUiCategory(decoratedCred.credentialType)

    try {
      categorizedClaims[uiCategory].push(decoratedCred)
    } catch (err) {
      categorizedClaims[uiCategory] = [decoratedCred]
    }
  })

  return categorizedClaims
}

// TODO Util, make subject mandatory
export const convertToDecoratedClaim = (
  vCreds: SignedCredential[],
): DecoratedClaims[] =>
  vCreds.map(vCred => {
    const claimData = { ...vCred.claim }
    delete claimData.id

    return {
      credentialType: getUiCredentialTypeByType(vCred.type),
      claimData,
      id: vCred.id,
      issuer: vCred.issuer,
      subject: vCred.claim.id || 'Not found',
      expires: vCred.expires || undefined,
    }
  })
