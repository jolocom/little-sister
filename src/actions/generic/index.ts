import { navigationActions } from 'src/actions/'
import { routeList } from 'src/routeList'
import SplashScreen from 'react-native-splash-screen'
import I18n from 'src/locales/i18n'
import { ThunkAction } from 'src/store'
import { AppError, ErrorCode } from 'src/lib/errors'
import settingKeys from '../../ui/settings/settingKeys'
import { SETTINGS } from '../../reducers/settings'

export const showErrorScreen = (
  error: AppError | Error,
): ThunkAction => dispatch => {
  const appError: AppError =
    error.constructor === AppError
      ? (error as AppError)
      : new AppError(ErrorCode.Unknown, error)

  return dispatch(
    navigationActions.navigate({
      routeName: routeList.Exception,
      params: {
        returnTo: appError.navigateTo,
        error: appError,
      },
    }),
  )
}

export const initApp: ThunkAction = async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  try {
    await backendMiddleware.initStorage()
    const storedSettings = await backendMiddleware.storageLib.get.settingsObject()
    // locale setup
    if (storedSettings.locale) I18n.locale = storedSettings.locale
    else storedSettings.locale = I18n.locale
    SplashScreen.hide()
    return dispatch(loadSettings(storedSettings))
  } catch (e) {
    return dispatch(
      showErrorScreen(
        new AppError(ErrorCode.WalletInitFailed, e, routeList.Landing),
      ),
    )
  }
}

export const loadSettings = (settings: { [key: string]: any }) => ({
  type: SETTINGS.LOAD_SETTINGS,
  value: settings,
})

export const setLocale = (locale: string): ThunkAction => async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  await backendMiddleware.storageLib.store.setting(settingKeys.locale, locale)
  I18n.locale = locale
  dispatch({ type: SETTINGS.SET_LOCALE, value: locale })

  // we need to reset the navigator so that all screens are re-rendered with the
  // new locale
  return dispatch(navigationActions.navigatorReset())
}
