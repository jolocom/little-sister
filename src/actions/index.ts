import * as accountActions from 'src/actions/account/'
import * as registrationActions from 'src/actions/registration/'
import * as navigationActions from 'src/actions/navigation/'
import * as genericActions from 'src/actions/generic/'
import * as ssoActions from 'src/actions/sso/'

export {
  accountActions,
  registrationActions,
  navigationActions,
  genericActions,
  ssoActions,
}
