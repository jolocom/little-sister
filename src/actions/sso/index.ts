import { Dispatch, AnyAction } from 'redux'
import { Linking } from 'react-native'
import { JolocomLib } from 'jolocom-lib'
import {
  StateCredentialRequestSummary,
  StateVerificationSummary,
} from 'src/reducers/sso'
import { BackendMiddleware } from 'src/backendMiddleware'
import { navigationActions, accountActions } from 'src/actions'
import { routeList } from 'src/routeList'
import { SignedCredential } from 'jolocom-lib/js/credentials/signedCredential/signedCredential'
import { showErrorScreen } from 'src/actions/generic'
import { getUiCredentialTypeByType } from 'src/lib/util'
import { InteractionType } from 'jolocom-lib/js/interactionTokens/types'
import { resetSelected } from '../account'
import { CredentialOffer } from 'jolocom-lib/js/interactionTokens/credentialOffer'
import { PaymentRequest } from 'jolocom-lib/js/interactionTokens/paymentRequest'
import { JSONWebToken } from 'jolocom-lib/js/interactionTokens/JSONWebToken'
import { CredentialsReceive } from 'jolocom-lib/js/interactionTokens/credentialsReceive'
import { CredentialRequest } from 'jolocom-lib/js/interactionTokens/credentialRequest'
import { getIssuerPublicKey } from 'jolocom-lib/js/utils/helper'
import { SoftwareKeyProvider } from 'jolocom-lib/js/vaultedKeyProvider/softwareProvider'
import { KeyTypes } from 'jolocom-lib/js/vaultedKeyProvider/types'
import { consumePaymentRequest } from './paymentRequest'
import { Authentication } from 'jolocom-lib/js/interactionTokens/authentication'
import { consumeAuthenticationRequest } from './authenticationRequest'

export const setCredentialRequest = (
  request: StateCredentialRequestSummary,
) => ({
  type: 'SET_CREDENTIAL_REQUEST',
  value: request,
})

export const clearInteractionRequest = () => ({
  type: 'CLEAR_INTERACTION_REQUEST',
})

export const setReceivingCredential = (external: SignedCredential[]) => ({
  type: 'SET_EXTERNAL',
  external,
})

export const setDeepLinkLoading = (value: boolean) => ({
  type: 'SET_DEEP_LINK_LOADING',
  value,
})

export const parseJWT = (encodedJwt: string) => async (
  dispatch: Dispatch<AnyAction>,
) => {
  dispatch(accountActions.toggleLoading(true))
  try {
    const returnedDecodedJwt = await JolocomLib.parse.interactionToken.fromJWT(
      encodedJwt,
    )

    switch (returnedDecodedJwt.interactionType) {
      case InteractionType.CredentialRequest:
        return dispatch(
          consumeCredentialRequest(returnedDecodedJwt as JSONWebToken<
            CredentialRequest
          >),
        )
      case InteractionType.CredentialOffer:
        return dispatch(
          consumeCredentialOfferRequest(returnedDecodedJwt as JSONWebToken<
            CredentialOffer
          >),
        )
      case InteractionType.CredentialsReceive:
        return dispatch(
          receiveExternalCredential(returnedDecodedJwt as JSONWebToken<
            CredentialsReceive
          >),
        )
      case InteractionType.PaymentRequest:
        return dispatch(
          consumePaymentRequest(returnedDecodedJwt as JSONWebToken<
            PaymentRequest
          >),
        )
      case InteractionType.Authentication:
        return dispatch(
          consumeAuthenticationRequest(returnedDecodedJwt as JSONWebToken<
            Authentication
          >),
        )
      default:
        return new Error('Unknown interaction type when parsing JWT')
    }
  } catch (err) {
    console.log('error: ', err)
    dispatch(accountActions.toggleLoading(false))
    dispatch(setDeepLinkLoading(false))
    dispatch(showErrorScreen(err))
  }
}

export const consumeCredentialOfferRequest = (
  credOfferRequest: JSONWebToken<CredentialOffer>,
) => async (
  dispatch: Dispatch<AnyAction>,
  getState: Function,
  backendMiddleware: BackendMiddleware,
) => {
  const { keyChainLib, identityWallet, registry } = backendMiddleware

  try {
    await identityWallet.validateJWT(credOfferRequest, undefined, registry)

    const password = await keyChainLib.getPassword()
    const credOfferResponse = await identityWallet.create.interactionTokens.response.offer(
      {
        callbackURL: credOfferRequest.interactionToken.callbackURL,
        instant: credOfferRequest.interactionToken.instant,
        requestedInput: {},
      },
      password,
      credOfferRequest,
    )

    const res = await fetch(credOfferRequest.interactionToken.callbackURL, {
      method: 'POST',
      body: JSON.stringify({ token: credOfferResponse.encode() }),
      headers: { 'Content-Type': 'application/json' },
    }).then(body => body.json())

    dispatch(parseJWT(res.token))
  } catch (err) {
    dispatch(accountActions.toggleLoading(false))
    dispatch(setDeepLinkLoading(false))
    dispatch(showErrorScreen(new Error('JWT Token parse failed')))
  }
}

export const receiveExternalCredential = (
  credReceive: JSONWebToken<CredentialsReceive>,
) => async (
  dispatch: Dispatch<AnyAction>,
  getState: Function,
  backendMiddleware: BackendMiddleware,
) => {
  const { identityWallet, registry } = backendMiddleware

  try {
    await identityWallet.validateJWT(credReceive, undefined, registry)
  } catch (error) {
    console.log(error)
    dispatch(
      showErrorScreen(
        new Error('Validation of external credential token failed'),
      ),
    )
  }

  try {
    const providedCredentials = credReceive.interactionToken.signedCredentials

    const results = await Promise.all(
      providedCredentials.map(async vcred => {
        const remoteIdentity = await registry.resolve(vcred.issuer)
        return SoftwareKeyProvider.verifyDigestable(
          getIssuerPublicKey(vcred.signer.keyId, remoteIdentity.didDocument),
          vcred,
        )
      }),
    )

    if (results.every(el => el === true)) {
      dispatch(setReceivingCredential(providedCredentials))
      dispatch(accountActions.toggleLoading(false))
      dispatch(
        navigationActions.navigatorReset({
          routeName: routeList.CredentialDialog,
        }),
      )
    } else {
      dispatch(accountActions.toggleLoading(false))
      dispatch(showErrorScreen(new Error('Signature validation failed')))
    }
  } catch (error) {
    dispatch(accountActions.toggleLoading(false))
    dispatch(
      showErrorScreen(
        new Error('Signature validation on external credential failed'),
      ),
    )
  }
}

interface AttributeSummary {
  type: string[]
  results: Array<{
    verification: string
    fieldName: string
    values: string[]
  }>
}

export const consumeCredentialRequest = (
  decodedCredentialRequest: JSONWebToken<CredentialRequest>,
) => async (
  dispatch: Dispatch<AnyAction>,
  getState: Function,
  backendMiddleware: BackendMiddleware,
) => {
  const { storageLib, identityWallet, registry } = backendMiddleware
  const { did } = getState().account.did.toJS()

  try {
    await identityWallet.validateJWT(
      decodedCredentialRequest,
      undefined,
      registry,
    )
    const requestedTypes =
      decodedCredentialRequest.interactionToken.requestedCredentialTypes
    const attributesForType = await Promise.all<AttributeSummary>(
      requestedTypes.map(storageLib.get.attributesByType),
    )

    const populatedWithCredentials = await Promise.all(
      attributesForType.map(async entry => {
        if (entry.results.length) {
          return Promise.all(
            entry.results.map(async result => ({
              type: getUiCredentialTypeByType(entry.type),
              values: result.values,
              verifications: await storageLib.get.verifiableCredential({
                id: result.verification,
              }),
            })),
          )
        }

        return [
          {
            type: getUiCredentialTypeByType(entry.type),
            values: [],
            verifications: [],
          },
        ]
      }),
    )

    const abbreviated = populatedWithCredentials.map(attribute =>
      attribute.map(entry => ({
        ...entry,
        verifications: entry.verifications.map((vCred: SignedCredential) => ({
          id: vCred.id,
          issuer: vCred.issuer,
          selfSigned: vCred.signer.did === did,
          expires: vCred.expires,
        })),
      })),
    )
    const flattened = abbreviated.reduce((acc, val) => acc.concat(val))

    // TODO requester shouldn't be optional
    const summary = {
      callbackURL: decodedCredentialRequest.interactionToken.callbackURL,
      requester: decodedCredentialRequest.issuer,
      availableCredentials: flattened,
      requestJWT: decodedCredentialRequest.encode(),
    }

    dispatch(setCredentialRequest(summary))
    dispatch(accountActions.toggleLoading(false))
    dispatch(navigationActions.navigatorReset({ routeName: routeList.Consent }))
    dispatch(setDeepLinkLoading(false))
  } catch (error) {
    console.log(error)
    dispatch(accountActions.toggleLoading(false))
    dispatch(
      showErrorScreen(new Error('Consumption of credential request failed')),
    )
  }
}

export const sendCredentialResponse = (
  selectedCredentials: StateVerificationSummary[],
) => async (
  dispatch: Dispatch<AnyAction>,
  getState: Function,
  backendMiddleware: BackendMiddleware,
) => {
  const { storageLib, keyChainLib, encryptionLib, registry } = backendMiddleware
  const { activeCredentialRequest } = getState().sso

  try {
    const password = await keyChainLib.getPassword()
    const decryptedSeed = encryptionLib.decryptWithPass({
      cipher: await storageLib.get.encryptedSeed(),
      pass: password,
    })
    const userVault = new SoftwareKeyProvider(
      Buffer.from(decryptedSeed, 'hex'),
      password,
    )

    const wallet = await registry.authenticate(userVault, {
      derivationPath: KeyTypes.jolocomIdentityKey,
      encryptionPass: password,
    })

    const credentials = await Promise.all(
      selectedCredentials.map(
        async cred =>
          (await storageLib.get.verifiableCredential({ id: cred.id }))[0],
      ),
    )

    const jsonCredentials = credentials.map(cred => cred.toJSON())

    const request = JolocomLib.parse.interactionToken.fromJWT(
      activeCredentialRequest.requestJWT,
    )
    const credentialResponse = await wallet.create.interactionTokens.response.share(
      {
        callbackURL: activeCredentialRequest.callbackURL,
        suppliedCredentials: jsonCredentials,
      },
      password,
      request,
    )

    if (activeCredentialRequest.callbackURL.includes('http')) {
      await fetch(activeCredentialRequest.callbackURL, {
        method: 'POST',
        body: JSON.stringify({ token: credentialResponse.encode() }),
        headers: { 'Content-Type': 'application/json' },
      })
    } else {
      const url =
        activeCredentialRequest.callbackURL + credentialResponse.encode()
      Linking.openURL(url)
    }
    dispatch(cancelSSO())
  } catch (error) {
    // TODO: better error message
    console.log(error)
    dispatch(accountActions.toggleLoading(false))
    dispatch(
      showErrorScreen(
        new Error('The credential response could not be created'),
      ),
    )
  }
}

export const cancelSSO = () => (dispatch: Dispatch<AnyAction>) => {
  dispatch(clearInteractionRequest())
  dispatch(accountActions.toggleLoading(false))
  dispatch(navigationActions.navigatorReset({ routeName: routeList.Home }))
}

export const cancelReceiving = () => (dispatch: Dispatch<AnyAction>) => {
  dispatch(resetSelected())
  dispatch(navigationActions.navigatorReset({ routeName: routeList.Home }))
}
