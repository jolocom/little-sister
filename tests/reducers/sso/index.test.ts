import { ssoReducer } from 'src/reducers/sso'
import { ssoActions as actions } from 'src/actions/'

describe('sso reducer', () => {
  it('should initialize correctly', () => {
    expect(ssoReducer(undefined, { type: '@INIT' })).toMatchSnapshot()
  })

  it('should handle the SET_CREDENTIAL_REQUEST action ', () => {
    expect(
      ssoReducer(
        undefined,
        actions.setCredentialRequest({
          requester: 'did:jolo:mock',
          callbackUrl: 'http://example.com/myendpoint/',
          request: [
            {
              type: ['Credential', 'ProofOfEmailCredential'],
              constraints: [],
            },
          ],
        }),
      ),
    ).toMatchSnapshot()
  })

  it('should handle the CLEAR_CREDENTIAL_REQUEST action ', () => {
    expect(
      ssoReducer(undefined, actions.setCredentialRequest()),
    ).toMatchSnapshot()
  })
})
