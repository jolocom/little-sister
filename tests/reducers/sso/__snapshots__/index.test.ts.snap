// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sso reducer should handle the CLEAR_CREDENTIAL_REQUEST action  1`] = `
Object {
  "activeAuthenticationRequest": Object {
    "callbackURL": "",
    "description": "",
    "requestJWT": "",
    "requester": "",
  },
  "activeCredentialRequest": undefined,
  "activePaymentRequest": Object {
    "amount": 0,
    "callbackURL": "",
    "description": "",
    "paymentRequest": "",
    "receiver": Object {
      "address": "",
      "did": "",
    },
  },
  "deepLinkLoading": false,
}
`;

exports[`sso reducer should handle the SET_CREDENTIAL_REQUEST action  1`] = `
Object {
  "activeAuthenticationRequest": Object {
    "callbackURL": "",
    "description": "",
    "requestJWT": "",
    "requester": "",
  },
  "activeCredentialRequest": Object {
    "callbackUrl": "http://example.com/myendpoint/",
    "request": Array [
      Object {
        "constraints": Array [],
        "type": Array [
          "Credential",
          "ProofOfEmailCredential",
        ],
      },
    ],
    "requester": "did:jolo:mock",
  },
  "activePaymentRequest": Object {
    "amount": 0,
    "callbackURL": "",
    "description": "",
    "paymentRequest": "",
    "receiver": Object {
      "address": "",
      "did": "",
    },
  },
  "deepLinkLoading": false,
}
`;

exports[`sso reducer should initialize correctly 1`] = `
Object {
  "activeAuthenticationRequest": Object {
    "callbackURL": "",
    "description": "",
    "requestJWT": "",
    "requester": "",
  },
  "activeCredentialRequest": Object {
    "availableCredentials": Array [],
    "callbackURL": "",
    "requestJWT": "",
    "requester": "",
  },
  "activePaymentRequest": Object {
    "amount": 0,
    "callbackURL": "",
    "description": "",
    "paymentRequest": "",
    "receiver": Object {
      "address": "",
      "did": "",
    },
  },
  "deepLinkLoading": false,
}
`;
